File: crates/spfs/spfs/tracking/_manifest_test.py
from typing import Callable
from collections import OrderedDict
import io
import os
import json
import random

import py.path
import pytest

from .. import encoding
from ._manifest import (
    ManifestBuilder,
    Manifest,
    Entry,
    EntryKind,
    compute_tree,
    compute_entry,
    compute_manifest,
    layer_manifests,
)
from ._diff import compute_diff


def test_compute_tree_determinism() -> None:

    first = compute_tree("./spfs")
    second = compute_tree("./spfs")
    assert first == second


def test_compute_manifest() -> None:

    root = os.path.abspath("./spfs")
    this = os.path.relpath(__file__, root)
    manifest = compute_manifest(root)
    assert manifest.get_path(this) is not None


def test_manifest_relative_paths(tmpdir: py.path.local) -> None:

    tmpdir.join("dir1.0/dir2.0/file.txt").write("somedata", ensure=True)
    tmpdir.join("dir1.0/dir2.1/file.txt").write("someotherdata", ensure=True)
    tmpdir.join("dir2.0/file.txt").write("evenmoredata", ensure=True)
    tmpdir.join("a_file.txt").write("rootdata", ensure=True)

    manifest = compute_manifest(tmpdir.strpath)
    with pytest.raises(FileNotFoundError):
        # should be no entry for root - as there is not enough info
        # about the root to form an entry (missing mode and name)
        manifest.get_path("/")
        # but we should still be able to list the entries of root
    assert manifest.list_dir("/"), "should be able to list root"
    assert manifest.get_path("/dir1.0/dir2.0/file.txt") is not None
    assert manifest.get_path("dir1.0/dir2.1/file.txt") is not None


def test_manifest_sorting(tmpdir: py.path.local) -> None:

    tmpdir = tmpdir.join("data")
    tmpdir.join("dir1.0/dir2.0/file.txt").write("somedata", ensure=True)
    tmpdir.join("dir1.0/dir2.1/file.txt").write("someotherdata", ensure=True)
    tmpdir.join("dir1.0/file.txt").write("thebestdata", ensure=True)
    tmpdir.join("dir2.0/file.txt").write("evenmoredata", ensure=True)
    tmpdir.join("a_file.txt").write("rootdata", ensure=True)
    tmpdir.join("z_file.txt").write("rootdata", ensure=True)

    manifest = ManifestBuilder(tmpdir.strpath)
    compute_entry(tmpdir.strpath, append_to=manifest)

    final = manifest.finalize()
    actual = list(p for p, _ in final.walk())
    expected = [
        "/a_file.txt",
        "/dir1.0",
        "/dir1.0/dir2.0",
        "/dir1.0/dir2.0/file.txt",
        "/dir1.0/dir2.1",
        "/dir1.0/dir2.1/file.txt",
        "/dir1.0/file.txt",
        "/dir2.0",
        "/dir2.0/file.txt",
        "/z_file.txt",
    ]
    assert actual == expected


def test_layer_manifests(tmpdir: py.path.local) -> None:

    a_dir = tmpdir.join("a").ensure(dir=True)
    a_dir.join("a.txt").write("a", ensure=True)
    a_dir.join("both.txt").write("a", ensure=True)
    a = compute_manifest(a_dir.strpath)

    b_dir = tmpdir.join("b").ensure(dir=True)
    b_dir.join("b.txt").write("b", ensure=True)
    b_dir.join("both.txt").write("b", ensure=True)
    b = compute_manifest(b_dir.strpath)

    both_dir = tmpdir.join("both").ensure(dir=True)
    both_dir.join("a.txt").write("a", ensure=True)
    both_dir.join("b.txt").write("b", ensure=True)
    both_dir.join("both.txt").write("b", ensure=True)
    both = compute_manifest(both_dir.strpath)

    actual = layer_manifests(a, b)

    assert actual.digest() == both.digest()


def test_layer_manifests_removal() -> None:

    a = ManifestBuilder("/")
    a.add_entry(
        "/a_only",
        Entry(
            kind=EntryKind.BLOB,
            mode=0o000777,
            name="a_only",
            object=encoding.EMPTY_DIGEST,
            size=0,
        ),
    )

    b = ManifestBuilder("/")
    b.add_entry(
        "/a_only",
        Entry(
            kind=EntryKind.MASK,
            mode=0o020000,
            name="a_only",
            object=encoding.EMPTY_DIGEST,
            size=0,
        ),
    )

    actual = layer_manifests(b.finalize(), a.finalize(), b.finalize())
    entry = actual.get_path("/a_only")
    assert entry is not None
    assert entry.kind is EntryKind.MASK


def test_manifest_builder_remove_file() -> None:

    builder = ManifestBuilder("/")
    builder.add_entry(
        "/entry",
        Entry(
            kind=EntryKind.BLOB,
            mode=0o000777,
            name="entry",
            object=encoding.EMPTY_DIGEST,
            size=0,
        ),
    )
    builder.remove_entry("/entry")

    manifest = builder.finalize()
    with pytest.raises(FileNotFoundError):
        manifest.get_path("/entry")
    with pytest.raises(FileNotFoundError):
        manifest.get_path("entry")


def test_manifest_builder_remove_dir() -> None:

    builder = ManifestBuilder("/")
    builder.add_entry(
        "/entry",
        Entry(
            kind=EntryKind.TREE,
            mode=0o000777,
            name="entry",
            object=encoding.EMPTY_DIGEST,
            size=0,
        ),
    )
    builder.remove_entry("/entry")

    manifest = builder.finalize()
    with pytest.raises(FileNotFoundError):
        manifest.get_path("/entry")
    with pytest.raises(FileNotFoundError):
        manifest.get_path("entry")


repo_manifest = compute_manifest(".")


def test_manifest_layering_speed(benchmark: Callable) -> None:

    benchmark(layer_manifests, repo_manifest, repo_manifest)


def test_manifest_encoding_speed(benchmark: Callable) -> None:

    repo_manifest = compute_manifest(".")
    stream = io.BytesIO()

    @benchmark
    def encode_decode() -> None:
        stream.seek(0)
        repo_manifest.encode(stream)
        stream.seek(0)
        Manifest.decode(stream)

File: crates/spfs/src/tracking/manifest_test.rs
// Copyright (c) 2021 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use rstest::rstest;

use super::{compute_manifest, EntryKind, Manifest};
use crate::graph;

fixtures!();

#[rstest]
#[tokio::test]
async fn test_compute_manifest_determinism() {
    let first = compute_manifest("./src").await.unwrap();
    let second = compute_manifest("./src").await.unwrap();
    assert_eq!(first, second);
}

#[rstest]
#[tokio::test]
async fn test_compute_manifest() {
    let root = std::fs::canonicalize("./src").unwrap();
    let this = file!().to_string().replace("./", "").replace("src/", "");
    let manifest = compute_manifest(root).await.unwrap();
    assert!(manifest.get_path(&this).is_some());
}

#[rstest]
#[tokio::test]
async fn test_manifest_relative_paths(tmpdir: tempdir::TempDir) {
    let dir = tmpdir.path();
    ensure(dir.join("dir1.0/dir2.0/file.txt"), "somedata");
    ensure(dir.join("dir1.0/dir2.1/file.txt"), "someotherdata");
    ensure(dir.join("dir2.0/file.txt"), "evenmoredata");
    ensure(dir.join("a_file.txt"), "rootdata");

    let manifest = compute_manifest(dir).await.unwrap();
    assert!(
        manifest.list_dir("/").is_some(),
        "should be able to list root"
    );
    assert!(manifest.get_path("/dir1.0/dir2.0/file.txt").is_some());
    assert!(manifest.get_path("dir1.0/dir2.1/file.txt").is_some());
}
#[rstest]
#[tokio::test]
async fn test_manifest_sorting(tmpdir: tempdir::TempDir) {
    let dir = tmpdir.path().join("data");
    ensure(dir.join("dir1.0/dir2.0/file.txt"), "somedata");
    ensure(dir.join("dir1.0/dir2.1/file.txt"), "someotherdata");
    ensure(dir.join("dir1.0/file.txt"), "thebestdata");
    ensure(dir.join("dir2.0/file.txt"), "evenmoredata");
    ensure(dir.join("a_file.txt"), "rootdata");
    ensure(dir.join("z_file.txt"), "rootdata");

    let manifest = compute_manifest(dir).await.unwrap();

    let mut actual: Vec<_> = manifest.walk().collect();
    actual.sort();
    let actual: Vec<_> = actual.into_iter().map(|n| n.path).collect();
    let expected = vec![
        "/dir1.0",
        "/dir1.0/dir2.0",
        "/dir1.0/dir2.0/file.txt",
        "/dir1.0/dir2.1",
        "/dir1.0/dir2.1/file.txt",
        "/dir1.0/file.txt",
        "/dir2.0",
        "/dir2.0/file.txt",
        "/a_file.txt",
        "/z_file.txt",
    ];
    assert_eq!(actual, expected);
}
#[rstest]
#[tokio::test]
async fn test_layer_manifests(tmpdir: tempdir::TempDir) {
    let a_dir = tmpdir.path().join("a");
    ensure(a_dir.join("a.txt"), "a");
    ensure(a_dir.join("both.txt"), "a");
    let mut a = compute_manifest(a_dir).await.unwrap();

    let b_dir = tmpdir.path().join("b");
    ensure(b_dir.join("b.txt"), "b");
    ensure(b_dir.join("both.txt"), "b");
    let b = compute_manifest(b_dir).await.unwrap();

    let both_dir = tmpdir.path().join("both");
    ensure(both_dir.join("a.txt"), "a");
    ensure(both_dir.join("b.txt"), "b");
    ensure(both_dir.join("both.txt"), "b");
    let both = compute_manifest(both_dir).await.unwrap();

    a.update(&b);

    assert_eq!(a, both);
    assert_eq!(graph::Manifest::from(&a), graph::Manifest::from(&both));
}
#[rstest]
#[tokio::test]
async fn test_layer_manifests_removal() {
    let mut a = Manifest::default();
    a.mkfile("a_only").unwrap();

    let mut b = Manifest::default();
    let mut node = b.mkfile("a_only").unwrap();
    node.kind = EntryKind::Mask;

    let mut c = Manifest::default();
    c.update(&a);
    assert!(c.get_path("/a_only").unwrap().kind.is_blob());
    c.update(&b);
    assert!(c.get_path("/a_only").unwrap().kind.is_mask());

    compute_manifest("./src").await.unwrap();
}
